### PR TITLE
Fix spec declarations, add return spec instrumentation, tidy up core_test.clj. test with Opensearch

### PR DESCRIPTION
**Pull request by jdt-at-litera**
TL;DR: make spec declarations work with return-validating instrumentation support Opensearch in the core_test execution, fix test pre-existing test case bugs, and tidy up some of the setup logic in the test.

It probably should have been separate commits (one for specs, one for tests), sorry 'bout that.

1. Corrects some bugs in clojure specs for return values of some functions. This is necessary for organizations enabling return spec validation if they want to call spandex APIs from their own tests.  The alternative is to write instrumentation logic that avoids instrumenting the spandex namespaces.

2. Adds 'orchestra' to dependencies and enables its instrumentation so that clojure.spec return values are routinely checked in spandex unit tests. Note that clojure.test.spec.alpha's instrumentation does not check return values.

3. Cleans up the core_test.clj module to to perform some test setup in a fixture instead of in a top level form in the module at load/compile-time.

   I left the connection client initialization as a top level form for now in case it's useful for debugging client interactions in the REPL, though it isn't a particularly tidy thing to do.

4. Validates the core_test.clj module against an Opensearch container instead of Elasticsearch, in order ensure at least minimal Opensearch compatibility.

   Note that while Opensearch forked Elasticsearch at 7.10 or so, Opensearch contines on its own somewhat divergent path separate from Elasticsearch and it's hard to say what sorts of complications may preclude future use of Spandex with Opensearch.  That said, Opensearch 2.0.1 service works with Spandex as of January 1, 2023 (with the changes in this PR - AFAIK - sniffing not verified).

5. Change use of depreciated Elasticsearch APIs to use their replacements. Spandex is currently testing against containers running Elasticsearch 7.6, which happily takes syntax which has since been deprecated later in the 7.x branch sequence.

   Elasticsearch 7.15 deprecates POST syntax of `/index/type/doc-id` in favor of `/index/_doc/doc-id`.  For details see: https://www.elastic.co/guide/en/elasticsearch/reference/7.15/removal-of-types.html

   Similarly `/index/type/_search` is now `/index/_search` for scrolling tests.

   While deprecated constructs only generate warnings in later versions of Elasticsearch, it errors completely out in more recent versions of Opensearch.

6. Fixes broken core_test.clj test cases that couldn't possibly have run succesfully via `lein test`, AFAICT. The two broken tests are:

   `test-head-req` which is doing a HEAD on a document that doesn't exist, and getting a 404 when it doesn't seem to expect a 404.  I've changed the test to create a document before doing a HEAD on it.

   `test-scrolling-chan` was broken because it was posting 99 documents and then asserting that 100 documents were read. I've changed the assertion to be 99.

7. Replaced the `wait!` function that did a `Thread/sleep` with a `POST /index/_refresh` which is what I assume `wait!` was waiting for.  Thread/sleep is never a good synchronization mechanism.

I verified tests with Elasticsearch 7.6.1 and Opensearch 2.0.1 containers. Everything worked with 7.6.1, the version used in the github ci setup. Everything worked, with security warnings, with ES 7.17.8.

Sniffing (and only sniffing) fails with Opensearch 2.0.1.  Whether it's a problem with container host mapping setup or Opensearch specific issues I don't know.  I haven't attempted to fix this problem yet.

I was unable to figure out how to eliminate sufficient security parameters on my Opensearch container so that I could use the default client configuration in the unit test. The client configuration used for Opensearch is commented out.

### CODE REVIEW

**Comment by mpenet** [comment]
Thanks. I will have a look at this pr on Monday first thing.

**Comment by mpenet** [comment]
That's looking good, my only request would be to put orchestra in dev dependencies.

**Comment by jdt-at-litera** [comment]
I’ll take a look.

Dave Tenny
Senior Staff Software Engineer

Litera.com<https://www.litera.com>

**Comment by jdt-at-litera** [comment]
> That's looking good, my only request would be to put orchestra in dev dependencies.

Done

**Comment by mpenet** [comment]
Merged, thanks
I ll do a release later this afternoon.

**Comment by jdt-at-litera** [comment]
Great, here's hoping I didn't screw anything up!
